src/python/pants/engine/internals/native.py

```diff
@@ -31,38 +48,74 @@
 from pants.util.memo import memoized_property
 from pants.util.meta import SingletonMetaclass
 
+RuleYield = Union[Get, Tuple[Get, ...]]
+RuleSend = TypeVar("RuleSend", Tuple[Any, ...], Exception)
+RuleReturn = Any
+RuleCoroutine = Coroutine[RuleYield, RuleSend, RuleReturn]
+
+EngineGeneratorResult = Union[
+    PyGeneratorResponseGet,
+    PyGeneratorResponseGetMulti,
+    PyGeneratorResponseBreak,
+    PyGeneratorResponseThrow,
+]
+
+
 logger = logging.getLogger(__name__)
 
 
+class IntrinsicError(Exception):
+    """Exceptions raised within intrinsic methods implemented in Rust."""
+
+
 class Externs:
     """Methods exposed from Python to Rust.
 
     TODO: These could be implemented in Rust in `externs.rs` via the cpython API.
     """
 
-    def __init__(self, lib):
+    def __init__(self, lib: ModuleType):
         self.lib = lib
 
     _do_raise_keyboardinterrupt = bool(os.environ.get("_RAISE_KEYBOARDINTERRUPT_IN_EXTERNS", False))
 
-    def is_union(self, input_type):
+    def is_union(self, input_type: Type) -> bool:
         """Return whether or not a type is a member of a union."""
         # NB: This check is exposed for testing error handling in CFFI methods. This code path should
         # never be active in normal pants usage.
         return union.is_instance(input_type)
 
-    def create_exception(self, msg):
+    def create_exception(self, msg: str) -> IntrinsicError:
         """Given a utf8 message string, create an Exception object."""
-        return Exception(msg)
+        return IntrinsicError(msg)
+
+    @staticmethod
+    def _do_send(func: RuleCoroutine, arg: Tuple[Any, ...]) -> RuleYield:
+        return func.send(arg)
+
+    def generator_send(self, func: RuleCoroutine, arg: Tuple[Any, ...]) -> EngineGeneratorResult:
+        return self._generator_send_or_throw_impl(func, arg, self._do_send)
 
-    def generator_send(
-        self, func, arg
-    ) -> Union[PyGeneratorResponseGet, PyGeneratorResponseGetMulti, PyGeneratorResponseBreak]:
+    @staticmethod
+    def _do_throw(func: RuleCoroutine, arg: Exception) -> RuleYield:
+        # Coroutine.throw(self, typ, val=None, tb=None) will raise `typ` (the first positional
+        # argument) as the exception if `val` and `tb` are None (as they are here).
+        return func.throw(arg)  # type: ignore[arg-type]
+
+    def generator_throw(self, func: RuleCoroutine, arg: Exception) -> EngineGeneratorResult:
+        return self._generator_send_or_throw_impl(func, arg, self._do_throw)
+
+    def _generator_send_or_throw_impl(
```

Comment:
I don't have a concrete suggestion here, but I have a gut feeling that the logic around using one of `generator_send` or `generator_throw` to call into `_generator_send_or_throw_impl`  could possibly be simplified. `RuleCoroutine` seems like a rather complicated type definition, and I wonder if it's possible to adjust how the rust code invokes these two functions so that we don't need to define it.

Reply:
I have slimmed down the type definitions, removed the separate `generator_throw()` method, and moved the types closer to this function definition. Let me know if this seems to make sense.